### ISSUE TITLE
Pod `do-not-consolidate` annotation to allow Nodes to be replaced for reasons other than consolidation

### ISSUE THREAD

**Issue by si-c613** [aws/karpenter#4756]
### Description

**What problem are you trying to solve?**

We have some workloads in our cluster where disruption is OK, but less desirable. Things like CoreDNS, Ingress, Prometheus scrapers. Karpenter allows our clusters to be highly elastic and mixed with our HPAs Karpenter is often replacing Nodes as part of consolidation. 

It is desirable that these workloads are left alone during consolidation but the Node is replaced for other interruptions such as TTL expiry or drift detection.

Currently we can add the annotation `karpenter.sh/do-not-evict: "true"` to stop these being evicted, however it is my understanding that this would block the Nodes from being replaced entirely and could lead to these being the only Pods running on the Node.

Nodes have the `karpenter.sh/do-not-consolidate: "true"` annotation, I wonder if there could be a way to apply a similar annotation to Pods to allow them to be evicted for drift/expiry but not for consolidation.

**How important is this feature to you?**

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by engedaam** [comment]
Thank you for the issue! The team is currently working option for users to have more control over derpovisioing. Here is a mega issue on the topic: https://github.com/aws/karpenter/issues/1738. I believe [machine disruption gate](https://github.com/aws/karpenter/issues/1302) would fit your needs. There is a deign proposed and I think your use case can give good insight https://github.com/aws/karpenter-core/pull/516

**Comment by engedaam** [comment]
If [machine disruption gate](https://github.com/aws/karpenter/issues/1302) fits your use case, we can close this issue

**Comment by si-c613** [comment]
@engedaam  thanks for the reply, I've had a read through some of what look like the 'key comments' but there's a lot in there so apologies if I've missed something! I do like the idea of maintenance windows and Node Disruption Budgets as a concept, they'd be a great addition, but I don't think they'd help with the situation here. as we still want consolidation to happen.

Hopefully the below helps illustrate the intent rather than add confusion!

Assuming it's our ingress we want less disruption on, we have two identically sized Nodes with room for 4 Pods. A HPA has caused the below spread of Pods due to an app2 scaling down.
| Node1 | Node 2 |
|--------|--------|
| ingress | app1|
| | app1 |
| | app2 |
| | |

There are two possibilities here, Karpenter can either drain `Node1` and  the ingress Pod moves to `Node2` or Karpenter could drain `Node2` and the app Pods move to `Node1`.

Currently we can do this by adding the `karpenter.sh/do-not-evict: "true"` annotation to the ingress Pod, however my understanding is that this would block any TTL or drift actions.

With maintenance gates I believe we'd have the same problem.

The thought is if we have a Pod level `karpenter.sh/do-not-consolidate: "true"` annotation that could block `Node1` from being drained due to consolidation but still allow TTL/drift actions to happen, potentially during a maintenance window.

**Comment by njtran** [comment]
So you're saying that you're fine with disruption to your ingress app if it's for the purpose of a node TTL/Drift, but not in the case of Consolidation? Is there a time where you don't want any disruption, including TTL/Drift and Consolidation to occur? If so, how do you model the your tolerance to disruptions there? Can you model your disruptions to treat all disruptions the same?

**Comment by si-c613** [comment]
> So you're saying that you're fine with disruption to your ingress app if it's for the purpose of a node TTL/Drift, but not in the case of Consolidation? 

Yes this is correct, but this is just one example of where/how this might be used.

> Is there a time where you don't want any disruption, including TTL/Drift and Consolidation to occur?

Maintenance gates would cover that I think. I don't think I spotted it in the other issue but how do these handle Pods with `karpenter.sh/do-not-evict: "true"` ?

>  If so, how do you model the your tolerance to disruptions there?

So I think maintenance gates would cover that off, but also with the above question how do we enable maintenance in the maintenance window, still enable consolidation but also still protect Jobs or other workloads we want to finish before the maintenance action? 
It feels a bit like with maintenance windows we'd either be in the same place regarding consolidation or still blocking the maintenance.

> Can you model your disruptions to treat all disruptions the same?

I think I get what you're asking here but apologies if I'm answering a different question.
Bit of a tangent, I've used Ingress in the example above as we know we have an issue with Keep Alive terminations not happening correctly which we believe to be outside of the Ingress Pods. We have an idea of what we need to do to solve this by moving TLS into the Pods rather than the NLB, but this means we have challenges around generating certificates and a few other challenges that take time and need scheduling around other priorities. 
With this in mind it is a lot easier for us to stick an annotation onto the Pods to stop them being evicted, but in this case that means our Nodes are drifting and need intervention to keep them up to date.

I feel there is a difference between these being regularly churned due to the lean clusters Karpenter allows us to run (including use of spot) and a semi-regular bit of churn due to Node updates which in combination with maintenance gates would allow us to schedule the changes at a lower traffic period of time.
